<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Open-the-data-file-and-have-a-look-at-the-general-information." data-toc-modified-id="Open-the-data-file-and-have-a-look-at-the-general-information.-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Open the data file and have a look at the general information.</a></span></li><li><span><a href="#Data-exploration" data-toc-modified-id="Data-exploration-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Data exploration</a></span></li><li><span><a href="#Data-transformation" data-toc-modified-id="Data-transformation-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Data transformation</a></span><ul class="toc-item"><li><span><a href="#Investigate-income_type" data-toc-modified-id="Investigate-income_type-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Investigate <code>income_type</code></a></span></li><li><span><a href="#Restoring-missing-values-in-total_income" data-toc-modified-id="Restoring-missing-values-in-total_income-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Restoring missing values in <code>total_income</code></a></span></li><li><span><a href="#Restoring-values-in-days_employed" data-toc-modified-id="Restoring-values-in-days_employed-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Restoring values in <code>days_employed</code></a></span></li></ul></li><li><span><a href="#Categorization-of-data" data-toc-modified-id="Categorization-of-data-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Categorization of data</a></span></li><li><span><a href="#Checking-the-Hypotheses" data-toc-modified-id="Checking-the-Hypotheses-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Checking the Hypotheses</a></span></li><li><span><a href="#Reviewer's-Summary" data-toc-modified-id="Reviewer's-Summary-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Reviewer's Summary</a></span></li></ul></div>

# Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building the **credit score** of a potential customer. The **credit score** is used to evaluate the ability of a potential borrower to repay their loan.


## Open the data file and have a look at the general information. 


In [518]:
# Loading all the libraries
import pandas as pd


In [519]:
df = pd.read_csv('/datasets/credit_scoring_eng.csv')

In [520]:
# Load the data
df

children  days_employed  dob_years            education  education_id  \
0             1   -8437.673028         42    bachelor's degree             0   
1             1   -4024.803754         36  secondary education             1   
2             0   -5623.422610         33  Secondary Education             1   
3             3   -4124.747207         32  secondary education             1   
4             0  340266.072047         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1   -4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1   -2113.346888         38  secondary education             1   
21523         3   -3112.481705         38  secondary education             1   
21524         2   -1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose  
0         40620.102    purchase of the house  
1         17932.802             car purchase  
2         23341.752    purchase of the house  
3         42820.568  supplementary education  
4         25378.572        to have a wedding  
...             ...                      ...  
21520     35966.698     housing transactions  
21521     24959.969        purchase of a car  
21522     14347.610                 property  
21523     39054.888        buying my own car  
21524     13127.587             to buy a car  

[21525 rows x 12 columns]

In [521]:
print(df.head())

   children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house 

In [522]:
display(df.head())

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

In [523]:
df.head()

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

## Data exploration

In [524]:
# Let's see how many rows and columns our dataset has
print(df.shape)


(21525, 12)


There are 21,525 row and 12 columns.

In [525]:
# let's print the first N rows
df.head(20)


children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42    bachelor's degree             0   
1          1   -4024.803754         36  secondary education             1   
2          0   -5623.422610         33  Secondary Education             1   
3          3   -4124.747207         32  secondary education             1   
4          0  340266.072047         53  secondary education             1   
5          0    -926.185831         27    bachelor's degree             0   
6          0   -2879.202052         43    bachelor's degree             0   
7          0    -152.779569         50  SECONDARY EDUCATION             1   
8          2   -6929.865299         35    BACHELOR'S DEGREE             0   
9          0   -2188.756445         41  secondary education             1   
10         2   -4171.483647         36    bachelor's degree             0   
11         0    -792.701887         40  secondary education             1   
12         0            NaN         65  secondary education             1   
13         0   -1846.641941         54         some college             2   
14         0   -1844.956182         56    bachelor's degree             0   
15         1    -972.364419         26  secondary education             1   
16         0   -1719.934226         35  secondary education             1   
17         0   -2369.999720         33    bachelor's degree             0   
18         0  400281.136913         53  secondary education             1   
19         0  -10038.818549         48  SECONDARY EDUCATION             1   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
12  civil partnership                 1      M     retiree     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   
15            married                 0      F    employee     0   
16            married                 0      F    employee     0   
17  civil partnership                 1      M    employee     0   
18    widow / widower                 2      F     retiree     0   
19           divorced                 3      F    employee     0   

    total_income                              purpose  
0      40620.102                purchase of the house  
1      17932.802                         car purchase  
2      23341.752                purchase of the house  
3      42820.568              supplementary education  
4      25378.572                    to have a wedding  
5      40922.170                purchase of the house  
6      38484.156                 housing transactions  
7      21731.829                            education  
8      15337.093                     having a wedding  
9      23108.150  purchase of the house for my family  
10     18230.959                      buy real estate  
11     12331.077           buy commercial real estate  
12           NaN                    to have a wedding  
13     20873.317                         car purchase  
14     26420.466          buy residential real estate  
15     18691.345         construction of own property  
16   

1. There are negative values for days_employed.
2. 'dob_years' column name is confusing. Shall be using 'age'.
3. There are duplicate values in education, eg. 'bachelor's degree' & 'BACHELOR'S DEGREE'.
4. No standardisation of purpose, eg. 'supplementary education', 'education'.
5. There are some empty cells.
6. There are missing values.

In [526]:
# Get info on data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Yes, there are missing values in columns 'days_employed & 'total_income', which is not the same as RangeIndex of 21525 entries.

In [527]:
# Let's look in the filtered table at the the first column with missing data
print(df.isna().sum())

missing_employed_data = df[df['days_employed'].isnull()]
missing_employed_data.head(10)


children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64


children  days_employed  dob_years            education  education_id  \
12         0            NaN         65  secondary education             1   
26         0            NaN         41  secondary education             1   
29         0            NaN         63  secondary education             1   
41         0            NaN         50  secondary education             1   
55         0            NaN         54  secondary education             1   
65         0            NaN         21  secondary education             1   
67         0            NaN         52    bachelor's degree             0   
72         1            NaN         32    bachelor's degree             0   
82         2            NaN         50    bachelor's degree             0   
83         0            NaN         52  secondary education             1   

        family_status  family_status_id gender    income_type  debt  \
12  civil partnership                 1      M        retiree     0   
26            married                 0      M  civil servant     0   
29          unmarried                 4      F        retiree     0   
41            married                 0      F  civil servant     0   
55  civil partnership                 1      F        retiree     1   
65          unmarried                 4      M       business     0   
67            married                 0      F        retiree     0   
72            married                 0      M  civil servant     0   
82            married                 0      F       employee     0   
83            married                 0      M       employee     0   

    total_income                                   purpose  
12           NaN                         to have a wedding  
26           NaN                                 education  
29           NaN                    building a real estate  
41           NaN                  second-hand car purchase  
55           NaN                         to have a wedding  
65           NaN  transactions with commercial real estate  
67           NaN       purchase of the house for my family  
72           NaN  transactions with commercial real estate  
82           NaN                                   housing  
83           NaN                                   housing

There are missing values in 2 columns, and the counts of missing values in these 2 columns are the same (2174).

Since the missing values are consistent between these 2 columns, it suggests that the missing samples are of the same size. Therefore, the missing values in these columns may be considered symmetric.


In [528]:
# Let's apply multiple conditions for filtering data and look at the number of rows in the filtered table.
filtered_data = df[(df['days_employed'].isna()) & (df['total_income'].isna())]
filtered_data


children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

**Intermediate conclusion**

The number of rows in the filtered table match the number of missing values. It might be that the data for 'days_employed' & 'total_income' is not provided/keyed in by users.


The percentage of the missing values compared to the whole set is around 10% of missing values compared to whole dataset (divide 2174 by 21525). The missing values in 'days_employed' & 'total_income' might be just left out by the clients while they are providing the information or they do not want to disclose.


We need to find the mean/median to fill in the missing values.

In [529]:
# Let's investigate clients who do not have data on identified characteristic and the column with the missing values
df_na = df[(df['days_employed'].isna()) & (df['total_income'].isna())]
df_na.groupby('income_type').count()

children  days_employed  dob_years  education  education_id  \
income_type                                                                  
business            508              0        508        508           508   
civil servant       147              0        147        147           147   
employee           1105              0       1105       1105          1105   
entrepreneur          1              0          1          1             1   
retiree             413              0        413        413           413   

               family_status  family_status_id  gender  debt  total_income  \
income_type                                                                  
business                 508               508     508   508             0   
civil servant            147               147     147   147             0   
employee                1105              1105    1105  1105             0   
entrepreneur               1                 1       1     1             0   
retiree                  413               413     413   413             0   

               purpose  
income_type             
business           508  
civil servant      147  
employee          1105  
entrepreneur         1  
retiree            413

In [530]:
# Checking distribution
display(df_na.describe())


children  days_employed    dob_years  education_id  \
count  2174.000000            0.0  2174.000000   2174.000000   
mean      0.552438            NaN    43.632015      0.800828   
std       1.469356            NaN    12.531481      0.530157   
min      -1.000000            NaN     0.000000      0.000000   
25%       0.000000            NaN    34.000000      0.250000   
50%       0.000000            NaN    43.000000      1.000000   
75%       1.000000            NaN    54.000000      1.000000   
max      20.000000            NaN    73.000000      3.000000   

       family_status_id         debt  total_income  
count       2174.000000  2174.000000           0.0  
mean           0.975161     0.078197           NaN  
std            1.418220     0.268543           NaN  
min            0.000000     0.000000           NaN  
25%            0.000000     0.000000           NaN  
50%            0.000000     0.000000           NaN  
75%            1.000000     0.000000           NaN  
max            4.000000     1.000000           NaN

In the missing_data table, for those rows with missing values of 'days_employed' & 'total_income:
1. The min value for 'children' is -1, which is not logic.
2. 'dob_years' minimum is 0, which is not logic too.

**Possible reasons for missing values in data**

The missing values in 'days_employed' & 'total_income' is just random, and might be just left out by the clients while they are providing the information or they do not want to disclose. In addition to that, there are some others information is missing from the missing rows/data.



In [531]:
# Checking the distribution in the whole dataset
display(df.describe())

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

**Intermediate conclusion**

The distribution in the original dataset is not similar to the distribution of the filtered table. As there are missing informaiton in the missing values rows, so the means are not the same.


Despite of the above finding, we still can't make conclusion yet. Furhter investigation of dataset is needed.

In [532]:
# Check for other reasons and patterns that could lead to missing values



**Intermediate conclusion**

The missing values shall be not completed while gathering data.
Other reasons for missing values: might be key in error/human error

In [533]:
# Checking for other patterns - explain which


In the missing_data table, for those rows with missing values of 'days_employed' & 'total_income:
1. The min value for 'children' is -1, which is not logic.
2. 'dob_years' minimum is 0, which is not logic too.

**Conclusions**

In the missing_data table, for those rows with missing values of 'days_employed' & 'total_income:
1. The min value for 'children' is -1, which is not logic.
2. 'dob_years' & 'debt' minimum is 0, which is not logic too.

The missing values shall be replaced using mean/median.

Data transformation is needed:
1. check and replace the missing values with mean/median.
2. check for duplicates and if to drop duplicates too.
3. check for different registers and incorrect artifacts and try to fix them if possible.

## Data transformation

In [534]:
# Let's see all values in education column to check if and what spellings will need to be fixed

# There are duplicated entry with different cases. Need to apply 'str.lower()' to standardise.
print(df['education'].unique())

["bachelor's degree" 'secondary education' 'Secondary Education'
 'SECONDARY EDUCATION' "BACHELOR'S DEGREE" 'some college'
 'primary education' "Bachelor's Degree" 'SOME COLLEGE' 'Some College'
 'PRIMARY EDUCATION' 'Primary Education' 'Graduate Degree'
 'GRADUATE DEGREE' 'graduate degree']


In [535]:
# Fix the registers if required
df['education'] = df['education'].str.lower()

In [536]:
# Checking all the values in the column to make sure we fixed them
print(df['education'].unique())

["bachelor's degree" 'secondary education' 'some college'
 'primary education' 'graduate degree']


[Check the data the `children` column]

In [537]:
# Let's see the distribution of values in the `children` column

display(df['children'].describe())
print('\n')
print(df['children'].unique())
print('\n')
print(df[df['children'] == -1].count())
print('\n')
print(df[df['children'] == 20].count())
print('\n')
print(df['children'].count())

count    21525.000000
mean         0.538908
std          1.381587
min         -1.000000
25%          0.000000
50%          0.000000
75%          1.000000
max         20.000000
Name: children, dtype: float64



[ 1  0  3  2 -1  4 20  5]


children            47
days_employed       44
dob_years           47
education           47
education_id        47
family_status       47
family_status_id    47
gender              47
income_type         47
debt                47
total_income        44
purpose             47
dtype: int64


children            76
days_employed       67
dob_years           76
education           76
education_id        76
family_status       76
family_status_id    76
gender              76
income_type         76
debt                76
total_income        67
purpose             76
dtype: int64


21525


There are value of -1 (47 rows) and max 20 (76 rows) children numbers which are not logic.
% of problematic data = (47+76)/21525 * 100% = 0.6%.

Will replace the problematic data with medain/mean.

In [538]:
# [fix the data based on your decision]
children_mean = df["children"].mean()
df["children"].replace(-1, children_mean, inplace=True)
df["children"].replace(20, children_mean, inplace=True)
df["children"] = df["children"].astype(int)
display(df["children"].describe())

count    21525.000000
mean         0.470476
std          0.750534
min          0.000000
25%          0.000000
50%          0.000000
75%          1.000000
max          5.000000
Name: children, dtype: float64

In [539]:
# Checking the `children` column again to make sure it's all fixed
print(df['children'].unique())

[1 0 3 2 4 5]


In [540]:
# Find problematic data in `days_employed`, if they exist, and calculate the percentage

print(df['days_employed'].describe())
print('\n')
df_days_employed_negative = df[(df['days_employed'] < 0)]
print(df_days_employed_negative.count())
print('\n')
df_days_employed_outlier = df[(df['days_employed'] > 100000)]
print('\n')
print(df_days_employed_outlier.count())
display(df_days_employed_outlier.head(20))
print('\n')
percentage_problematic_data = len(df_days_employed_negative + df_days_employed_outlier) / len(df) * 100
print(percentage_problematic_data)

count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64


children            15906
days_employed       15906
dob_years           15906
education           15906
education_id        15906
family_status       15906
family_status_id    15906
gender              15906
income_type         15906
debt                15906
total_income        15906
purpose             15906
dtype: int64




children            3445
days_employed       3445
dob_years           3445
education           3445
education_id        3445
family_status       3445
family_status_id    3445
gender              3445
income_type         3445
debt                3445
total_income        3445
purpose             3445
dtype: int64


children  days_employed  dob_years            education  education_id  \
4           0  340266.072047         53  secondary education             1   
18          0  400281.136913         53  secondary education             1   
24          1  338551.952911         57  secondary education             1   
25          0  363548.489348         67  secondary education             1   
30          1  335581.668515         62  secondary education             1   
35          0  394021.072184         68  secondary education             1   
50          0  353731.432338         63  secondary education             1   
56          0  370145.087237         64  secondary education             1   
71          0  338113.529892         62  secondary education             1   
78          0  359722.945074         61    bachelor's degree             0   
86          0  383933.549664         64  secondary education             1   
87          0  337659.535004         53  secondary education             1   
88          0  373129.789654         63    bachelor's degree             0   
98          0  364906.205736         54    bachelor's degree             0   
99          0  346541.618895          0  secondary education             1   
101         0  391558.961849         65  secondary education             1   
104         0  366323.649429         59  secondary education             1   
119         0  382189.129056         60  secondary education             1   
128         0  379035.920619         67  secondary education             1   
129         0  361010.321350         61  secondary education             1   

         family_status  family_status_id gender income_type  debt  \
4    civil partnership                 1      F     retiree     0   
18     widow / widower                 2      F     retiree     0   
24           unmarried                 4      F     retiree     0   
25             married                 0      M     retiree     0   
30             married                 0      F     retiree     0   
35   civil partnership                 1      M     retiree     0   
50             married                 0      F     retiree     0   
56     widow / widower                 2      F     retiree     0   
71             married                 0      F     retiree     0   
78             married                 0      M     retiree     0   
86   civil partnership                 1      M     retiree     0   
87             married                 0      F     retiree     0   
88   civil partnership                 1      F     retiree     0   
98             married                 0      F     retiree     0   
99             married                 0      F     retiree     0   
101            married                 0      F     retiree     0   
104          unmarried                 4      F     retiree     0   
119          unmarried                 4      F     retiree     0   
128           divorced                 3      F     retiree     0   
129  civil partnership                 1      M     retiree     0   

     total_income                                   purpose  
4       25378.572                         to have a wedding  
18       9091.804                  buying a second-hand car  
24      46487.558  transactions with commercial real estate  
25       8818.041                           buy real estate  
30      27432.971  transactions with commercial real estate  
35      12448.908                          having a wedding  
50      14774.837                                      cars  
56      23862.567                                 education  
71       7028.751                                      cars  
78      28020.423                         purchase of a car  
86      38885.353                          wedding ceremony  
87      11898.217           buying property for renting out  
88      33099.576                          having a wedding  
98      31953.168           buying property for 



89.90011614401858


In [541]:
# Address the problematic values, if they exist
# There are 15809 rows that the numbers in 'days_employed' columns are smaller than 0. And 3431 rows that the number is 'days_employed' columns are bigger than
10000.

df['days_employed'] = df['days_employed'].apply(abs)

# Calculate the median value of 'days_employed'
median_days_employed = df['days_employed'].median()

# Replace values greater than 100,000 (outlier) with the median value
df.loc[df['days_employed'] > 100000, 'days_employed'] = median_days_employed

print(df['days_employed'].describe())


count    19351.000000
mean      2324.746073
std       2089.961651
min         24.141633
25%        927.009265
50%       2194.220567
75%       2747.423625
max      18388.949901
Name: days_employed, dtype: float64


In [542]:
# Check the result - make sure it's fixed

df_days_employed_negative = df[(df['days_employed'] < 0)]
print(df_days_employed_negative.count())
print('\n')
df_days_employed_outlier = df[(df['days_employed'] > 100000)]
print('\n')
print(df_days_employed_outlier.count())
display(df_days_employed_outlier.head(20))
print('\n')
percentage_problematic_data = len(df_days_employed_negative + df_days_employed_outlier) / len(df) * 100
print(percentage_problematic_data)

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64




children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64


Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []



0.0


In [543]:
# Check the `dob_years` for suspicious values and count the percentage
display(df['dob_years'].describe())

problematic_dob_years = df[df['dob_years'] == 0]
print(problematic_dob_years)

percentage_of_problematic_dob_years = len(problematic_dob_years) / len(df) *100
print(percentage_of_problematic_dob_years)


count    21525.000000
mean        43.293380
std         12.574584
min          0.000000
25%         33.000000
50%         42.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

       children  days_employed  dob_years            education  education_id  \
99            0    2194.220567          0  secondary education             1   
149           0    2664.273168          0  secondary education             1   
270           3    1872.663186          0  secondary education             1   
578           0    2194.220567          0  secondary education             1   
1040          0    1158.029561          0    bachelor's degree             0   
...         ...            ...        ...                  ...           ...   
19829         0            NaN          0  secondary education             1   
20462         0    2194.220567          0  secondary education             1   
20577         0    2194.220567          0  secondary education             1   
21179         2     108.967042          0    bachelor's degree             0   
21313         0    1268.487728          0  secondary education             1   

      family_status  family_status_id g

There are 101 records that 'dob_years' is zero value, which does not make sense. These 101 records is only 0.5% of total dataset. Thus we could ignore this 101 rows with 'dob_years' is zero value.

In [544]:
# Address the issues in the `dob_years` column, if they exist
# The age in the `dob_years` column shall not be zero.

In [545]:
# Check the result - make sure it's fixed
df = df[df['dob_years'] != 0]
display(df['dob_years'].describe())

count    21424.000000
mean        43.497479
std         12.246934
min         19.000000
25%         33.000000
50%         43.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

[Now let's check the `family_status` column. See what kind of values there are and what problems you may need to address.]

In [546]:
# Let's see the values for the column
print(df['family_status'].unique())

['married' 'civil partnership' 'widow / widower' 'divorced' 'unmarried']


In [547]:
# Address the problematic values in `family_status`, if they exist
# no problem with the data in this column


In [548]:
# Check the result - make sure it's fixed


[Now let's check the `gender` column. See what kind of values there are and what problems you may need to address]

In [549]:
# Let's see the values in the column
print(df['gender'].unique())
display(df[df['gender'] == 'XNA'])


['F' 'M' 'XNA']


children  days_employed  dob_years     education  education_id  \
10701         0    2358.600502         24  some college             2   

           family_status  family_status_id gender income_type  debt  \
10701  civil partnership                 1    XNA    business     0   

       total_income          purpose  
10701     32624.825  buy real estate

In [550]:
# Address the problematic values, if they exist
# Shall be either 'F' or 'M', only 1 row with 'XNA' value, shall omit this record.

In [551]:
# Check the result - make sure it's fixed
df = df[df['gender'] != 'XNA']
print(df['gender'].unique())

['F' 'M']


[Now let's check the `income_type` column. See what kind of values there are and what problems you may need to address]

In [552]:
# Let's see the values in the column
print(df['income_type'].unique())

['employee' 'retiree' 'business' 'civil servant' 'unemployed'
 'entrepreneur' 'student' 'paternity / maternity leave']


In [553]:
# Address the problematic values, if they exist

In [554]:
# Check the result - make sure it's fixed


[Now let's see if we have any duplicates in our data. If we do, you'll need to decide what you will do with them and explain why.]

In [555]:
# Checking duplicates
display(df.describe())
print('\n')
df.duplicated().sum()


children  days_employed     dob_years  education_id  \
count  21423.000000   19259.000000  21423.000000  21423.000000   
mean       0.470476    2325.337406     43.498390      0.817859   
std        0.750348    2090.610992     12.246495      0.548178   
min        0.000000      24.141633     19.000000      0.000000   
25%        0.000000     926.823974     33.000000      1.000000   
50%        0.000000    2194.220567     43.000000      1.000000   
75%        1.000000    2748.436594     53.000000      1.000000   
max        5.000000   18388.949901     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21423.000000  21423.000000   19259.000000  
mean           0.971293      0.080894   26794.133121  
std            1.419770      0.272679   16494.347322  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16494.864000  
50%            0.000000      0.000000   23200.877000  
75%            1.000000      0.000000   32536.052000  
max            4.000000      1.000000  362496.645000

71

In [556]:
# Address the duplicates, if they exist
duplicated_df = df[df.duplicated()]
duplicated_df

children  days_employed  dob_years            education  education_id  \
2849          0            NaN         41  secondary education             1   
3290          0            NaN         58  secondary education             1   
4182          1            NaN         34    bachelor's degree             0   
4851          0            NaN         60  secondary education             1   
5557          0            NaN         58  secondary education             1   
...         ...            ...        ...                  ...           ...   
20702         0            NaN         64  secondary education             1   
21032         0            NaN         60  secondary education             1   
21132         0            NaN         47  secondary education             1   
21281         1            NaN         30    bachelor's degree             0   
21415         0            NaN         54  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
2849             married                 0      F    employee     0   
3290   civil partnership                 1      F     retiree     0   
4182   civil partnership                 1      F    employee     0   
4851   civil partnership                 1      F     retiree     0   
5557   civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
20702            married                 0      F     retiree     0   
21032            married                 0      F     retiree     0   
21132            married                 0      F    employee     0   
21281            married                 0      F    employee     0   
21415            married                 0      F     retiree     0   

       total_income                              purpose  
2849            NaN  purchase of the house for my family  
3290            NaN                    to have a wedding  
4182            NaN                     wedding ceremony  
4851            NaN                     wedding ceremony  
5557            NaN                    to have a wedding  
...             ...                                  ...  
20702           NaN              supplementary education  
21032           NaN                   to become educated  
21132           NaN                   housing renovation  
21281           NaN           buy commercial real estate  
21415           NaN                 housing transactions  

[71 rows x 12 columns]

In [557]:
# Last check whether we have any duplicates
df = df.drop_duplicates()
df.duplicated().sum()

0

In [558]:
# Check the size of the dataset that you now have after your first manipulations with it
print(df.shape)

(21352, 12)


The new dataset size now is 21352 rows, which is around 99% of the original dataset of 21525 rows. 

# Working with missing values

'education' & 'family_status_ are provided with IDs. Thus, these 2 columns are chose to work iwth dictionaries.

In [559]:
# Find the dictionaries

education_dict = {
    0: "bachelor's degree",
    1: 'secondary education',
    2: 'some college',
    3: 'primary education',
    4: 'graduate degree'
}

family_status_dict = {
    0: 'married',
    1: 'civil partnership',
    2: 'widow / widower',
    3: 'divorced',
    4: 'unmarried'
}



# By working with these dictionaries, we can enhance the quality and clarity of our analysis and provide more meaningful insights to the bank's loan division regarding the impact of marital status and number of children on loan default risk.

### Restoring missing values in `total_income`

Missing values are in columns 'days_employed & 'total_income'. Can calculate the mean/median to replace the missing values.


In [560]:
# Let's write a function that calculates the age category
def calculate_age_category(age):
    if age >= 18 and age <= 30:
        return 'young'
    elif age <= 50:
        return 'middle-aged'
    elif age <= 65:
        return 'senior'
    else:
        return 'elderly'

In [561]:
# Test if the function works
customer_age = 42
age_category = calculate_age_category(customer_age)
print(age_category)

middle-aged


In [562]:
# Creating new column based on function
df['age_category'] = df['dob_years'].apply(calculate_age_category)


In [563]:
# Checking how values in the new column
display(df['age_category'])


0        middle-aged
1        middle-aged
2        middle-aged
3        middle-aged
4             senior
            ...     
21520    middle-aged
21521        elderly
21522    middle-aged
21523    middle-aged
21524    middle-aged
Name: age_category, Length: 21352, dtype: object

Income is normally depends on education level and/or working experience.

In [564]:
# Create a table without missing values and print a few of its rows to make sure it looks fine
df_without_missing_values = df.dropna()
df_without_missing_values.head(20)

children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42    bachelor's degree             0   
1          1    4024.803754         36  secondary education             1   
2          0    5623.422610         33  secondary education             1   
3          3    4124.747207         32  secondary education             1   
4          0    2194.220567         53  secondary education             1   
5          0     926.185831         27    bachelor's degree             0   
6          0    2879.202052         43    bachelor's degree             0   
7          0     152.779569         50  secondary education             1   
8          2    6929.865299         35    bachelor's degree             0   
9          0    2188.756445         41  secondary education             1   
10         2    4171.483647         36    bachelor's degree             0   
11         0     792.701887         40  secondary education             1   
13         0    1846.641941         54         some college             2   
14         0    1844.956182         56    bachelor's degree             0   
15         1     972.364419         26  secondary education             1   
16         0    1719.934226         35  secondary education             1   
17         0    2369.999720         33    bachelor's degree             0   
18         0    2194.220567         53  secondary education             1   
19         0   10038.818549         48  secondary education             1   
20         1    1311.604166         36  secondary education             1   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   
15            married                 0      F    employee     0   
16            married                 0      F    employee     0   
17  civil partnership                 1      M    employee     0   
18    widow / widower                 2      F     retiree     0   
19           divorced                 3      F    employee     0   
20            married                 0      M    employee     0   

    total_income                              purpose age_category  
0      40620.102                purchase of the house  middle-aged  
1      17932.802                         car purchase  middle-aged  
2      23341.752                purchase of the house  middle-aged  
3      42820.568              supplementary education  middle-aged  
4      25378.572                    to have a wedding       senior  
5      40922.170                purchase of the house        young  
6      38484.156                 housing transactions  middle-aged  
7      21731.829                            education  middle-aged  
8      15337.093                     having a wedding  middle-aged  
9      23108.150  purchase of the house for my family  middle-aged  
10     18230.959                      buy real estate  middle-aged  
11     12331.077           buy commercial real estate  middle-aged  
13     20873.317                         car purchase       

In [565]:
# Look at the mean values for income based on your identified factors

mean_income_by_family_status = df.groupby(['family_status'])['total_income'].mean()
display(mean_income_by_family_status)
print('\n')
mean_income_by_children = df.groupby(['children'])['total_income'].mean()
display(mean_income_by_children)
print('\n')
mean_income_by_age_category = df.groupby(['age_category'])['total_income'].mean()
display(mean_income_by_age_category)


family_status
civil partnership    26702.249322
divorced             27202.683563
married              27045.383530
unmarried            26943.601742
widow / widower      23006.808776
Name: total_income, dtype: float64

children
0    26425.873208
1    27405.559686
2    27489.198728
3    29366.910652
4    27289.829647
5    27268.847250
Name: total_income, dtype: float64

age_category
elderly        20804.461461
middle-aged    28383.188755
senior         25175.685834
young          25815.651899
Name: total_income, dtype: float64

In [566]:
# Look at the median values for income based on your identified factors
median_income_by_family_status = df.groupby(['family_status'])['total_income'].median()
display(median_income_by_family_status)
print('\n')
median_income_by_children = df.groupby(['children'])['total_income'].median()
display(median_income_by_children)
print('\n')
median_income_by_age_category = df.groupby(['age_category'])['total_income'].median()
display(median_income_by_age_category)


family_status
civil partnership    23195.6360
divorced             23584.9695
married              23377.7080
unmarried            23139.4040
widow / widower      20523.2670
Name: total_income, dtype: float64

children
0    23033.3300
1    23661.4030
2    23136.1155
3    25191.6190
4    24981.6340
5    29816.2255
Name: total_income, dtype: float64

age_category
elderly        17978.2310
middle-aged    24711.8435
senior         21650.9475
young          22955.4740
Name: total_income, dtype: float64

The mean/median calculation is compared between 'family_status', 'children' & 'age_category'. These are the 3 factors that will affect the credit core directly.


I will use median for the missing value, as the difference between the lowest & highest earning is quite big.

In [567]:
#  Write a function that we will use for filling in missing values

# Step 1: Create a dictionary containing the median values for 'total_income' based on 'age_category'
median_income_by_age_category_dict = median_income_by_age_category.to_dict()

# Step 2: Map 'age_category' to corresponding median 'total_income' value
df['median_income_by_age_category'] = df['age_category'].map(median_income_by_age_category_dict)

# Step 3: Replace missing 'total_income' values with the mapped median values
df.loc[df['total_income'].isnull(), 'total_income'] = df.loc[df['total_income'].isnull(), 'median_income_by_age_category']

# Drop the temporary column used for mapping
df.drop(columns=['median_income_by_age_category'], inplace=True)

df['total_income'] = df['total_income'].astype(int)
print(df['total_income'].describe())

count     21352.000000
mean      26454.023604
std       15707.066457
min        3306.000000
25%       17223.500000
50%       23226.500000
75%       31321.250000
max      362496.000000
Name: total_income, dtype: float64


In [568]:
# Check if it works
print(df['total_income'].describe())  

count     21352.000000
mean      26454.023604
std       15707.066457
min        3306.000000
25%       17223.500000
50%       23226.500000
75%       31321.250000
max      362496.000000
Name: total_income, dtype: float64


In [569]:
# Apply it to every row


In [570]:
# Check if we got any errors
print(df['total_income'].isna().sum())

0


[If you've came across errors in preparing the values for missing data, it probably means there's something special about the data for the category. Give it some thought - you may want to fix some things manually, if there's enough data to find medians/means.]


In [571]:
# Replacing missing values if there are any errors


[When you think you've finished with `total_income`, check that the total number of values in this column matches the number of values in other ones.]

In [572]:
# Checking the number of entries in the columns
df.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 21352 entries, 0 to 21524
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21352 non-null  int64  
 1   days_employed     19259 non-null  float64
 2   dob_years         21352 non-null  int64  
 3   education         21352 non-null  object 
 4   education_id      21352 non-null  int64  
 5   family_status     21352 non-null  object 
 6   family_status_id  21352 non-null  int64  
 7   gender            21352 non-null  object 
 8   income_type       21352 non-null  object 
 9   debt              21352 non-null  int64  
 10  total_income      21352 non-null  int64  
 11  purpose           21352 non-null  object 
 12  age_category      21352 non-null  object 
dtypes: float64(1), int64(6), object(6)
memory usage: 2.3+ MB


###  Restoring values in `days_employed`

In [573]:
# Distribution of `days_employed` medians based on your identified parameters
median_days_employed_by_family_status = df.groupby(['family_status'])['days_employed'].median()
display(median_days_employed_by_family_status)
print('\n')
median_days_employed_by_children = df.groupby(['children'])['days_employed'].median()
display(median_days_employed_by_children)
print('\n')
median_days_employed_by_age_category = df.groupby(['age_category'])['days_employed'].median()
display(median_days_employed_by_age_category)

family_status
civil partnership    1948.413899
divorced             2194.220567
married              2194.220567
unmarried            1460.777023
widow / widower      2194.220567
Name: days_employed, dtype: float64

children
0    2194.220567
1    1666.838773
2    1683.917531
3    1734.879824
4    1905.879025
5    1231.571486
Name: days_employed, dtype: float64

age_category
elderly        2194.220567
middle-aged    1865.986695
senior         2194.220567
young          1045.752744
Name: days_employed, dtype: float64

In [574]:
# Distribution of `days_employed` means based on your identified parameters
mean_days_employed_by_family_status = df.groupby(['family_status'])['days_employed'].mean()
display(mean_income_by_family_status)
print('\n')
mean_days_employed_by_children = df.groupby(['children'])['days_employed'].mean()
display(mean_income_by_children)
print('\n')
mean_days_employed_by_age_category = df.groupby(['age_category'])['days_employed'].mean()
display(mean_income_by_age_category)
print('\n')
print(df['days_employed'].describe())

family_status
civil partnership    26702.249322
divorced             27202.683563
married              27045.383530
unmarried            26943.601742
widow / widower      23006.808776
Name: total_income, dtype: float64

children
0    26425.873208
1    27405.559686
2    27489.198728
3    29366.910652
4    27289.829647
5    27268.847250
Name: total_income, dtype: float64

age_category
elderly        20804.461461
middle-aged    28383.188755
senior         25175.685834
young          25815.651899
Name: total_income, dtype: float64



count    19259.000000
mean      2325.337406
std       2090.610992
min         24.141633
25%        926.823974
50%       2194.220567
75%       2748.436594
max      18388.949901
Name: days_employed, dtype: float64


I am using median as the difference is smaller.

In [575]:
# Let's write a function that calculates means or medians (depending on your decision) based on your identified parameter

# Step 1: Create a dictionary containing the median values for 'days_employed' based on 'age_category'
median_days_employed_by_age_category_dict = median_days_employed_by_age_category.to_dict()

# Step 2: Map 'age_category' to corresponding median 'days_employed' value
df['median_days_employed_by_age_category'] = df['age_category'].map(median_days_employed_by_age_category_dict)

# Step 3: Replace missing 'days_employed' values with the mapped median values
df.loc[df['days_employed'].isnull(), 'days_employed'] = df.loc[df['days_employed'].isnull(), 'median_days_employed_by_age_category']

# Step 4: Replace 'days_employed' outliers (>5000) with the mapped median values
df.loc[df['days_employed'] > 10000, 'days_employed'] = df.loc[df['days_employed'] > 10000, 'median_days_employed_by_age_category']

# Drop the temporary column used for mapping
df.drop(columns=['median_days_employed_by_age_category'], inplace=True)

df['days_employed'] = df['days_employed'].astype(int)
print(df['days_employed'].describe())      

count    21352.000000
mean      2158.466842
std       1679.500893
min         24.000000
25%       1023.000000
50%       2020.000000
75%       2421.250000
max       9993.000000
Name: days_employed, dtype: float64


In [576]:
# Check that the function works
print(df['days_employed'].isna().sum())


0


In [577]:
# Apply function to the income_type


In [578]:
# Check if function worked



In [579]:
# Replacing missing values



[When you think you've finished with `total_income`, check that the total number of values in this column matches the number of values in other ones.]

In [580]:
# Check the entries in all columns - make sure we fixed all missing values
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21352 entries, 0 to 21524
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21352 non-null  int64 
 1   days_employed     21352 non-null  int64 
 2   dob_years         21352 non-null  int64 
 3   education         21352 non-null  object
 4   education_id      21352 non-null  int64 
 5   family_status     21352 non-null  object
 6   family_status_id  21352 non-null  int64 
 7   gender            21352 non-null  object
 8   income_type       21352 non-null  object
 9   debt              21352 non-null  int64 
 10  total_income      21352 non-null  int64 
 11  purpose           21352 non-null  object
 12  age_category      21352 non-null  object
dtypes: int64(7), object(6)
memory usage: 2.3+ MB


## Categorization of data

In [581]:
# Print the values for your selected data for categorization
print(df['children'])
print(df['education'])
print(df['family_status'])
print(df['age_category'])
print(df['total_income'])


0        1
1        1
2        0
3        3
4        0
        ..
21520    1
21521    0
21522    1
21523    3
21524    2
Name: children, Length: 21352, dtype: int64
0          bachelor's degree
1        secondary education
2        secondary education
3        secondary education
4        secondary education
                ...         
21520    secondary education
21521    secondary education
21522    secondary education
21523    secondary education
21524    secondary education
Name: education, Length: 21352, dtype: object
0                  married
1                  married
2                  married
3                  married
4        civil partnership
               ...        
21520    civil partnership
21521              married
21522    civil partnership
21523              married
21524              married
Name: family_status, Length: 21352, dtype: object
0        middle-aged
1        middle-aged
2        middle-aged
3        middle-aged
4             senior
            ...   

[Let's check unique values]

In [582]:
# Check the unique values

print(df['children'].unique())
print(df['education'].unique())
print(df['family_status'].unique())
print(df['age_category'].unique())
print(df['total_income'].unique())

[1 0 3 2 4 5]
["bachelor's degree" 'secondary education' 'some college'
 'primary education' 'graduate degree']
['married' 'civil partnership' 'widow / widower' 'divorced' 'unmarried']
['middle-aged' 'senior' 'young' 'elderly']
[40620 17932 23341 ... 28489 24618 35966]


The main groups identified based on unique values: 
children, education, family_status, age_category, total_income


In [583]:
# Let's write a function to categorize the data based on common topics



In [584]:
# Create a column with the categories and count the values for them



[If you decide to categorize the numerical data, you'll need to come up with the categories for it too.]

In [585]:
# Looking through all the numerical data in your selected column for categorization
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21352 entries, 0 to 21524
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21352 non-null  int64 
 1   days_employed     21352 non-null  int64 
 2   dob_years         21352 non-null  int64 
 3   education         21352 non-null  object
 4   education_id      21352 non-null  int64 
 5   family_status     21352 non-null  object
 6   family_status_id  21352 non-null  int64 
 7   gender            21352 non-null  object
 8   income_type       21352 non-null  object
 9   debt              21352 non-null  int64 
 10  total_income      21352 non-null  int64 
 11  purpose           21352 non-null  object
 12  age_category      21352 non-null  object
dtypes: int64(7), object(6)
memory usage: 2.3+ MB


In [586]:
# Getting summary statistics for the column

print(df['total_income'].describe())

count     21352.000000
mean      26454.023604
std       15707.066457
min        3306.000000
25%       17223.500000
50%       23226.500000
75%       31321.250000
max      362496.000000
Name: total_income, dtype: float64


[Decide what ranges you will use for grouping and explain why.]

In [587]:
# Creating function for categorizing into different numerical groups based on ranges

# Create a new column 'children_category' with categorized values
def categorize_children(num_children):
    if num_children == 0:
        return 'No Children'
    elif num_children == 1:
        return '1 Child'
    elif num_children == 2:
        return '2 Children'
    else:
        return '3+ Children'




# Create a new column 'total_income_category' with categorized values
def categorize_total_income(total_income):
    if total_income <=10000:
        return 'low income'
    elif total_income <= 25000:
        return 'medium income'
    elif total_income <= 50000:
        return 'high income'
    else:
        return 'very high income'


In [588]:
# Creating column with categories
df['children_category'] = df['children'].apply(categorize_children)
display(df['children_category'])

print('\n')


df['total_income_category'] = df['total_income'].apply(categorize_total_income)
display(df['total_income_category'])


0            1 Child
1            1 Child
2        No Children
3        3+ Children
4        No Children
            ...     
21520        1 Child
21521    No Children
21522        1 Child
21523    3+ Children
21524     2 Children
Name: children_category, Length: 21352, dtype: object

0          high income
1        medium income
2        medium income
3          high income
4          high income
             ...      
21520      high income
21521    medium income
21522    medium income
21523      high income
21524    medium income
Name: total_income_category, Length: 21352, dtype: object

In [589]:
# Count each categories values to see the distribution
display(df['children_category'].value_counts())
print('\n')
display(df['education'].value_counts())
print('\n')
display(df['family_status'].value_counts())
print('\n')
display(df['age_category'].value_counts())
print('\n')
display(df['total_income_category'].value_counts())
print('\n')
df.info()

No Children    14143
1 Child         4792
2 Children      2039
3+ Children      378
Name: children_category, dtype: int64

secondary education    15108
bachelor's degree       5215
some college             741
primary education        282
graduate degree            6
Name: education, dtype: int64

married              12290
civil partnership     4129
unmarried             2794
divorced              1185
widow / widower        954
Name: family_status, dtype: int64

middle-aged    10992
senior          5942
young           3716
elderly          702
Name: age_category, dtype: int64

medium income       11873
high income          7239
very high income     1318
low income            922
Name: total_income_category, dtype: int64



<class 'pandas.core.frame.DataFrame'>
Int64Index: 21352 entries, 0 to 21524
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   children               21352 non-null  int64 
 1   days_employed          21352 non-null  int64 
 2   dob_years              21352 non-null  int64 
 3   education              21352 non-null  object
 4   education_id           21352 non-null  int64 
 5   family_status          21352 non-null  object
 6   family_status_id       21352 non-null  int64 
 7   gender                 21352 non-null  object
 8   income_type            21352 non-null  object
 9   debt                   21352 non-null  int64 
 10  total_income           21352 non-null  int64 
 11  purpose                21352 non-null  object
 12  age_category           21352 non-null  object
 13  children_category      21352 non-null  object
 14  total_income_category  21352 non-null  object
dtypes: int64(7), obje

## Checking the Hypotheses


**Is there a correlation between having children and paying back on time?**

In [590]:
# Check the children data and paying back on time

correlation_children = df['children'].corr(df['debt'])
print(correlation_children)


0.02495347753710729


In [591]:
# Create a groupby object (i.e. children group) and aggregate the select data (i.e. debt) using the aggregation function mean.
children_default_rate = (
    df
    .groupby("children")
    .agg({"debt": "mean"})
    .reset_index()
    .rename(columns={"debt": "default_rate"})
)
# Calculate the percentage default
children_default_rate["percent_default"] = children_default_rate["default_rate"] * 100
display(children_default_rate)

def get_default_rate(dataframe, column):
    # Create a groupby object (i.e. children group) and aggregate the select data (i.e. debt) using the aggregation function mean.
    data = (
        dataframe
        .groupby(column)
        .agg({"debt": "mean"})
        .reset_index()
        .rename(columns={"debt": "default_rate"})
    )
    # Calculate the percentage default
    data["percent_default"] = (data["default_rate"] * 100).round(2)
    return data

# Example of the function call

get_default_rate(df, "children")

children  default_rate  percent_default
0         0      0.075444         7.544368
1         1      0.092028         9.202838
2         2      0.095145         9.514468
3         3      0.082317         8.231707
4         4      0.097561         9.756098
5         5      0.000000         0.000000

children  default_rate  percent_default
0         0      0.075444             7.54
1         1      0.092028             9.20
2         2      0.095145             9.51
3         3      0.082317             8.23
4         4      0.097561             9.76
5         5      0.000000             0.00

Thank you sir for your suggestion and sharing. I have new insight of coding for this part! 

**Conclusion**

The correlation coefficient is positive (0.025). It suggests a positive correlation (having more children may be associated with a higher likelihood of defaulting). If it's close to zero, it implies a weak or no correlation.

Correlation does not imply causation. Even if there is a correlation, it does not mean that having children directly causes defaulting on loans. Other factors may be at play, and further analysis and domain expertise are necessary to draw meaningful conclusions.

The more children the client has, the higher the default rate.

**Is there a correlation between family status and paying back on time?**

In [592]:
# Check the family status data and paying back on time



# Calculating default-rate based on family status

# Group the filtered data by 'family_status' and calculate the default rate for each group
default_rate_by_family_status_mean = df.groupby('family_status')['debt'].mean().reset_index()
default_rate_by_family_status_median = df.groupby('family_status')['debt'].median().reset_index()

# Calculate the percentage of default 
default_rate_by_family_status_mean['Default_Rate_Percentage'] = default_rate_by_family_status_mean['debt'] * 100
default_rate_by_family_status_median['Default_Rate_Percentage'] = default_rate_by_family_status_median['debt'] * 100

# Sort the data by default rate in descending order
default_rate_by_family_status_mean.sort_values(by='Default_Rate_Percentage', ascending=False, inplace=True)
default_rate_by_family_status_median.sort_values(by='Default_Rate_Percentage', ascending=False, inplace=True)


# Display the default rates based on the number of children with debt
print(default_rate_by_family_status_mean)
print('\n')
print(default_rate_by_family_status_median)

       family_status      debt  Default_Rate_Percentage
3          unmarried  0.097709                 9.770938
0  civil partnership  0.093485                 9.348511
2            married  0.075427                 7.542718
1           divorced  0.071730                 7.172996
4    widow / widower  0.064990                 6.498952


       family_status  debt  Default_Rate_Percentage
0  civil partnership     0                        0
1           divorced     0                        0
2            married     0                        0
3          unmarried     0                        0
4    widow / widower     0                        0


**Conclusion**

The highest default rate category is unmarried and the least is widow/widower. It seems like no significant correlation between family status and paying back on time.

**Is there a correlation between income level and paying back on time?**

In [593]:
# Check the income level data and paying back on time



# Calculating default-rate based on income level

# Calculate the default rate
default_rate_by_income_mean = df.groupby('total_income_category')['debt'].mean().reset_index()
default_rate_by_income_median = df.groupby('total_income_category')['debt'].median().reset_index()

# Calculate the percentage of default 
default_rate_by_income_mean['Default_Rate_Percentage'] = default_rate_by_income_mean['debt'] * 100
default_rate_by_income_median['Default_Rate_Percentage'] = default_rate_by_income_median['debt'] * 100

# Sort the data by default rate in descending order
default_rate_by_income_mean.sort_values(by='Default_Rate_Percentage', ascending=False, inplace=True)
default_rate_by_income_median.sort_values(by='Default_Rate_Percentage', ascending=False, inplace=True)


# Display the default rates based on the number of children with debt
print(default_rate_by_income_mean)
print('\n')
print(default_rate_by_income_median)


  total_income_category      debt  Default_Rate_Percentage
2         medium income  0.084477                 8.447739
0           high income  0.080122                 8.012156
3      very high income  0.069803                 6.980273
1            low income  0.062907                 6.290672


  total_income_category  debt  Default_Rate_Percentage
0           high income     0                        0
1            low income     0                        0
2         medium income     0                        0
3      very high income     0                        0


**Conclusion**

The default rate of medium income is the highest among all categories while the default rate for low income is the lowest among all.  It seems like no significant correlation between family status and paying back on time.

**How does credit purpose affect the default rate?**

In [594]:
# Check the percentages for default rate for each credit purpose and analyze them

# Calculate the default rate for each credit purpose
default_rate_by_credit_purpose_mean = df.groupby('purpose')['debt'].mean().reset_index()
default_rate_by_credit_purpose_median = df.groupby('purpose')['debt'].median().reset_index()


# Calculate the percentage of default for each credit purpose
default_rate_by_credit_purpose_mean['Default_Rate_Percentage'] = default_rate_by_credit_purpose_mean['debt'] * 100
default_rate_by_credit_purpose_median['Default_Rate_Percentage'] = default_rate_by_credit_purpose_median['debt'] * 100

# Sort the data by default rate in descending order
default_rate_by_credit_purpose_mean.sort_values(by='Default_Rate_Percentage', ascending=False, inplace=True)
default_rate_by_credit_purpose_median.sort_values(by='Default_Rate_Percentage', ascending=False, inplace=True)

# Display the default rates and percentages for each credit purpose
print(default_rate_by_credit_purpose_mean)
print('\n')
print(default_rate_by_credit_purpose_median)

                                     purpose      debt  \
31          to get a supplementary education  0.114865   
22                         purchase of a car  0.110132   
14                  getting higher education  0.108491   
27                  second-hand car purchase  0.105372   
20                         profile education  0.101149   
33                              to own a car  0.096842   
29                        to become educated  0.095823   
10                                      cars  0.092437   
30                              to buy a car  0.092275   
6                          buying my own car  0.091633   
9                               car purchase  0.091503   
36                      university education  0.089087   
0                        building a property  0.087237   
15                       going to university  0.087221   
8                                        car  0.085714   
13                      getting an education  0.083900   
16            

**Conclusion**

The top 10 of the default rate are related to education & car purchase.

# General Conclusion 

Conclusion: 
The children number and the total income might bring impact on whether they will default on a loan but not the marital status.

1. The data is being looked into and filled the missing values with median/mean.
2. The duplicates are removed.
3. The possible reason for missing and duplicated values might be due to human error.
4. Suggest to have better system to capture the raw data or to set a template or dropdown list (if it is keyed into Microsoft Excel file).

